# Job Market Data Pipeline

**Stage 2 - Analysis**

In [1]:
import pandas as pd
import numpy as np 

In [2]:
df = pd.read_csv("processed/cleaned_postings.csv")

**Stage 2 is the analysis stage, where I interpret and explore the cleaned dataset 
to uncover meaningful insights about the job market. The questions I aim to answer 
are: which data roles are most in demand, which locations have the highest hiring 
activity, and what the salary landscape looked like during 2023-2024. Additionally, 
I will analyse the split between remote and on-site roles and identify which 
companies are hiring the most.**

### Most In Demand Roles

In [3]:
print(df['title'].value_counts().head(10)) 

title
Sales Manager                      672
Customer Service Representative    371
Project Manager                    354
Administrative Assistant           254
Senior Accountant                  236
Executive Assistant                228
Salesperson                        210
Registered Nurse                   207
Staff Accountant                   200
Receptionist                       200
Name: count, dtype: int64


**Here, I have identified the top 10 most frequently listed job roles during the 2023-2024 period. The most in demand roles were Sales Managaer, Customer Service Representatives and Project Manager. In order to gain specific insights into data roles, I will need to filter the dataset for data related job titles in a later step.**

### Top Hiring Locations

In [4]:
print(df["location"].value_counts().head(10))

location
United States    8103
New York, NY     2756
Chicago, IL      1830
Houston, TX      1754
Dallas, TX       1378
Atlanta, GA      1358
Boston, MA       1169
Austin, TX       1077
Charlotte, NC    1069
Phoenix, AZ      1048
Name: count, dtype: int64


**Here I have identified the top 10 hiring locations using the value_counts() method. 
The top result, "United States", likely represents remote or nationwide roles that 
did not specify a city rather than a physical location. New York dominates the 
city specific listings with 2,756 postings, whilst Texas stands out as a major 
hiring hub, appearing three times across Houston, Dallas and Austin.**

### Salary Landscape ###

In [5]:
print("Average salary =", np.mean(df['normalized_salary']))
print("Median salary =", np.median(df['normalized_salary']))
print("Highest salary =", np.max(df['normalized_salary']))
print("Lowest salary =", np.min(df['normalized_salary']))

Average salary = 60139.42616582417
Median salary = 0.0
Highest salary = 535600000
Lowest salary = 0


In [6]:
df_salaries = df[(df['normalized_salary'] > 20000) & (df['normalized_salary'] < 200000)] 

In [7]:
print(df_salaries.shape)

(33622, 12)


In [8]:
print((df['normalized_salary'] == 0).sum())

87078


In [9]:
print("Average salary =", np.mean(df_salaries["normalized_salary"]))
print("Median salary =", np.median(df_salaries["normalized_salary"]))
print("Highest salary =", np.max(df_salaries["normalized_salary"]))
print("Lowest salary =", np.min(df_salaries["normalized_salary"]))

Average salary = 87562.29730533579
Median salary = 79040.0
Highest salary = 199950
Lowest salary = 20782


**Initially I used the normalized_salary column to calculate average, median, highest and lowest salaries. However the results were inaccurate — the median and lowest salaries were 0, indicating an excess of null values that had been filled with 0, and the highest salary was \$535,600,000 which is clearly an anomaly or data entry error. To address this I filtered the dataset to only include salaries between \\$20,000 and \\$200,000, removing outliers and keeping the data realistic. This left 33,752 valid salary rows. The filtered results show salaries ranging from \\$20,782 to \\$199,950, with an average of \\$87,562 and a median of \\$79,040. The \\$8,500 gap between the two suggests that higher paying roles are pulling the average upward.**

### Remote vs On-Site roles
**Here, I analyse the split between remote and on-site job postings across the dataset.**

In [10]:
print(df.columns)

Index(['title', 'company_name', 'location', 'formatted_work_type',
       'remote_allowed', 'formatted_experience_level', 'min_salary',
       'max_salary', 'normalized_salary', 'views', 'applies', 'listed_time'],
      dtype='object')


In [11]:
print(df["remote_allowed"].value_counts())

remote_allowed
False    107830
True      15216
Name: count, dtype: int64


**Using the value_counts() method on the remote_allowed column, I uncovered that 
107,830 roles (87.6%) are on-site, whilst 15,216 roles (12.4%) are remote. 
This is a notable finding — despite the shift towards remote working following 
COVID-19, the overwhelming majority of roles advertised during 2023-2024 still 
require physical presence.**

### Top Hiring Companies ###
**Here, I analyse the top hiring companies amongst the job postings in this period.**

In [24]:
filtered_df = df[df["company_name"] != "Unknown"]
print(filtered_df["company_name"].value_counts().head(10))

company_name
Liberty Healthcare and Rehabilitation Services    1085
The Job Network                                   1000
J. Galt                                            604
Lowe's Companies, Inc.                             527
TEKsystems                                         524
Ingersoll Rand                                     513
Capital One                                        495
Cogent Communications                              452
Insight Global                                     418
Dice                                               414
Name: count, dtype: int64


**Excluding "Unknown" values, Liberty Healthcare and Rehabilitation Services leads 
the top 10 with 1,085 job postings, reflecting the high demand in the healthcare 
sector. The Job Network and J. Galt, both recruitment agencies posting on behalf 
of multiple companies, follow in 2nd and 3rd place. This theme continues with 
TEKsystems, Insight Global and Dice also appearing in the top 10, all of which 
are staffing agencies. Capital One and Lowe's stand out as the only recognisable 
direct employers in the list, suggesting that a significant portion of hiring is 
conducted through recruitment agencies rather than directly.**

### Analysis Summary ###
**Within Stage 2, I uncovered several key insights from the cleaned LinkedIn job 
postings dataset. The most in demand roles were Sales Manager, Customer Service 
Representative and Project Manager, indicating the dataset spans all industries. 
New York dominated hiring locations with 2,756 postings, suggesting it remains 
the strongest job market in the US. The average salary across valid postings was 
\\$87,562 with a median of \\$79,040, with the gap between the two indicating higher 
paying roles pulling the average upward. 87.6% of roles were on-site whilst only 
12.4% were remote, suggesting that physical presence is still the norm despite 
the post-COVID shift towards remote working. Finally, Liberty Healthcare led 
company hiring, with the majority of the top 10 being recruitment agencies rather 
than direct employers.**